### PR TITLE
store/tikv: fix retry without limit when committing the primary

### PR DESCRIPTION
**Pull request by hhkbp2**
Hi,

This PR tries to fix the retry without limit when committing the primary. If error happens when committing the primary, it will return a predefined error as "result undetermined" and then drop the connection to the corresponding mysql client.

PTAL @disksing @tiancaiamao @zimulala @AndreMouche

### CODE REVIEW

File: terror/terror.go [hunk 1]
```diff
@@ -44,6 +45,12 @@ const (
 	CodeMissConnectionID ErrCode = iota + 1
 )
 
+// Special error codes.
+const (
+	// CodeResultUndetermined indicates the sql execution result is undetermined.
+	CodeResultUndetermined ErrCode = iota + 1
```
Comment (disksing) [inline terror/terror.go]: Does _CodeResultUndetermined_ has the same value compare to _CodeMissConnectionID_?

Comment (hhkbp2) [inline terror/terror.go]: address comment.

**Comment by hhkbp2** [comment]
PTAL @disksing @tiancaiamao @zimulala @AndreMouche

File: store/tikv/2pc.go [hunk 2]
```diff
@@ -422,6 +414,24 @@ func (c *twoPhaseCommitter) commitSingleBatch(bo *Backoffer, batch batchKeys) er
...
+func (c *twoPhaseCommitter) commitSingleBatch(bo *Backoffer, batch batchKeys) error {
+	// If we fail to receive response for the request that commits primary key, it will be undetermined whether this
+	// transaction has been successfully committed.
+	// Under this circumstance,  we can not declare the commit is complete (may lead to data lost), nor can we throw
+	// an error (may lead to the duplicated key error when upper level restarts the transaction). Currently the best
+	// solution is to populate this error and let upper layer drop the connection to the corresponding mysql client.
+	isPrimary := false
+	if bytes.Compare(batch.keys[0], c.primary()) == 0 {
+		isPrimary = true
+	}
```
Comment (disksing) [inline store/tikv/2pc.go]: `isPrimary := bytes.Compare(batch.keys[0], c.primary()) == 0`

File: store/tikv/2pc.go [hunk 3]
```diff
@@ -422,6 +414,24 @@ func (c *twoPhaseCommitter) commitSingleBatch(bo *Backoffer, batch batchKeys) er
...
+
+	if err := c.doCommitSingleBatch(bo, batch); err != nil && isPrimary {
+		// change the Cause of the error to be returned
+		return errors.Wrap(err, terror.ErrResultUndetermined)
+	}
+	return nil
```
Comment (disksing) [inline store/tikv/2pc.go]: When err != nil && !isPrimary, it returns nil instead of error?

File: terror/terror.go [hunk 4]
```diff
@@ -44,6 +45,12 @@ const (
 	CodeMissConnectionID ErrCode = iota + 1
 )
 
+// Special error codes.
+const (
+	// CodeResultUndetermined indicates the sql execution result is undetermined.
+	CodeResultUndetermined ErrCode = iota + 2
+)
```
Comment (disksing) [inline terror/terror.go]: Maybe it's better to put all error codes together and not to use iota.

File: store/tikv/2pc.go (hunk 3 above)
Comment (tiancaiamao) [inline store/tikv/2pc.go]: if it's canceled, will this return ErrResultUndetermined ?
how to distinguish between manual cancel and canceled caused by backoff timeout ?
@hhkbp2

Comment (hhkbp2) [inline store/tikv/2pc.go]: @disksing Typo. It should return err.

Comment (hhkbp2) [inline store/tikv/2pc.go]: @tiancaiamao The context for executing the tikv tasks is not related to the context of client connection. So the error handling here is not effected by manual cancellation of either query or connection.

**Comment by hhkbp2** [comment]
PTAL @disksing @tiancaiamao @zimulala @AndreMouche

**Comment by tiancaiamao** [comment]
LGTM

**Comment by disksing** [comment]
LGTM. Please fix CI before merge.

**Pull request merged**